File: web/public/codec.js
// Text <-> binary conversion, shared by the browser and the server.
//
// The encoding is a real, lossless UTF-8 bit encoding: every byte of the UTF-8
// representation of the text becomes exactly eight '0'/'1' characters, so
// decode(encode(text)) === text for any string, CJK and emoji included.

export const BITS_PER_BYTE = 8;

export class BinaryDecodeError extends Error {}

/** Encode text as a string of 0s and 1s (8-bit groups joined by `separator`). */
export function encode(text, separator = " ") {
  return Array.from(new TextEncoder().encode(text), (byte) =>
    byte.toString(2).padStart(BITS_PER_BYTE, "0"),
  ).join(separator);
}

/** Decode a string of 0s and 1s back into text. Whitespace is ignored. */
export function decode(bits) {
  const stripped = bits.replace(/\s+/g, "");
  if (stripped === "") return "";
  if (!/^[01]+$/.test(stripped)) {
    throw new BinaryDecodeError("input contains characters other than 0 and 1");
  }
  if (stripped.length % BITS_PER_BYTE !== 0) {
    throw new BinaryDecodeError(
      `bit length ${stripped.length} is not a multiple of ${BITS_PER_BYTE}`,
    );
  }
  const bytes = new Uint8Array(stripped.length / BITS_PER_BYTE);
  for (let i = 0; i < bytes.length; i++) {
    bytes[i] = parseInt(
      stripped.slice(i * BITS_PER_BYTE, (i + 1) * BITS_PER_BYTE),
      2,
    );
  }
  try {
    return new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch (cause) {
    throw new BinaryDecodeError(`bits are not valid UTF-8: ${cause.message}`);
  }
}

/** True if `text` is a well-formed binary message. */
export function looksLikeBinary(text) {
  const stripped = text.replace(/\s+/g, "");
  return (
    stripped.length > 0 &&
    /^[01]+$/.test(stripped) &&
    stripped.length % BITS_PER_BYTE === 0
  );
}

File: binary_ai/codec.py
"""Text <-> binary conversion.

The encoding is a real, lossless UTF-8 bit encoding: every byte of the UTF-8
representation of the text becomes exactly eight '0'/'1' characters.  Nothing
is abbreviated and nothing is dropped, so ``decode(encode(text)) == text`` for
any string, including CJK text and emoji.
"""

from __future__ import annotations

BITS_PER_BYTE = 8


class BinaryDecodeError(ValueError):
    """Raised when a string is not a valid binary-encoded message."""


def encode(text: str, separator: str = " ") -> str:
    """Encode ``text`` as a string of 0s and 1s.

    Each UTF-8 byte becomes an 8-bit group; groups are joined by ``separator``
    (use ``""`` for one unbroken bit string).

    >>> encode("Hi")
    '01001000 01101001'
    """
    return separator.join(format(byte, "08b") for byte in text.encode("utf-8"))


def decode(bits: str) -> str:
    """Decode a string of 0s and 1s back into text.

    Any whitespace between groups is ignored, so both ``"01001000 01101001"``
    and ``"0100100001101001"`` decode to ``"Hi"``.
    """
    stripped = "".join(bits.split())
    if not stripped:
        return ""
    if stripped.strip("01"):
        raise BinaryDecodeError("input contains characters other than 0 and 1")
    if len(stripped) % BITS_PER_BYTE:
        raise BinaryDecodeError(
            f"bit length {len(stripped)} is not a multiple of {BITS_PER_BYTE}"
        )
    payload = bytes(
        int(stripped[i : i + BITS_PER_BYTE], 2)
        for i in range(0, len(stripped), BITS_PER_BYTE)
    )
    try:
        return payload.decode("utf-8")
    except UnicodeDecodeError as exc:  # pragma: no cover - defensive
        raise BinaryDecodeError(f"bits are not valid UTF-8: {exc}") from exc


def looks_like_binary(text: str) -> bool:
    """Return True if ``text`` is a well-formed binary message."""
    stripped = "".join(text.split())
    return (
        bool(stripped)
        and not stripped.strip("01")
        and len(stripped) % BITS_PER_BYTE == 0
    )
